src/cluster_migrateslots.c

```diff
@@ -0,0 +1,1954 @@
+            slotExportBeginStreaming(job);
+        } else {
+            serverLog(LL_WARNING,
+                      "Child process failed to snapshot slot migration %s",
+                      job->description);
+            finishSlotMigrationJob(job, SLOT_MIGRATION_JOB_FAILED, "Failed to perform snapshot");
+        }
+    }
+}
+
+/* Returns 1 if the attached slot migration job for the client is ready to flush the data from the
+ * output buffer. Returns 0 otherwise. While this returns 0, replicated data for this slot migration
+ * will accumulate in the client output buffer. */
+int clusterSlotMigrationShouldInstallWriteHandler(client *c) {
+    slotMigrationJob *job = c->slot_migration_job;
+    if (!job || job->type != SLOT_MIGRATION_EXPORT) {
+        return 1;
+    }
+    return job->state != SLOT_EXPORT_SNAPSHOTTING;
+}
+
+/* Feed the slot export jobs with the given argv and argc from the replication
+ * log. Slot is pre-populated using the original command that was run. */
+void clusterFeedSlotExportJobs(int dbid, robj **argv, int argc, int slot) {
+    if (slot == -1) {
+        /* We can safely ignore any commands with no keys. This includes
+         * MULTI/EXEC. This isn't a problem since the entire slot migration is
+         * atomically visible and therefore transactions are redundant. */
+        return;
+    }
+
+    listIter li;
+    listNode *ln;
+    listRewind(server.cluster->slot_migration_jobs, &li);
+
+    /* Select cmd will be generated only if at least one client needs it. */
+    robj *selectcmd = NULL;
+    while ((ln = listNext(&li))) {
+        slotMigrationJob *job = (slotMigrationJob *)ln->value;
+        if (job->type != SLOT_MIGRATION_EXPORT) continue;
+        if (!job->client) continue;
+        if (!isSlotMigrationJobInProgress(job) || job->state < SLOT_EXPORT_SNAPSHOTTING) continue;
+        if (!isSlotInSlotRanges(slot, job->slot_ranges)) continue;
+
+        if (dbid != job->client->db->id) {
+            serverAssert(selectDb(job->client, dbid) == C_OK);
+            if (!selectcmd) {
+                selectcmd = generateSelectCommand(dbid);
+            }
+            addReply(job->client, selectcmd);
+        }
+
+        addReplyArrayLen(job->client, argc);
+        for (int i = 0; i < argc; i++) {
+            addReplyBulk(job->client, argv[i]);
+        }
+    }
+    if (selectcmd) {
+        decrRefCount(selectcmd);
+    }
+}
+
+int isSlotExportPauseTimedOut(slotMigrationJob *job) {
+    return job->mf_end < mstime() || !getPausedActionsWithPurpose(PAUSE_DURING_SLOT_MIGRATION);
+}
+
+/* Revalidate that the migration's slots are still owned by ourselves, or fail the migration
+ * otherwise. */
+void updateSlotExportIfOwnershipChanged(slotMigrationJob *job) {
+    serverAssert(job->type == SLOT_MIGRATION_EXPORT);
+    if (!isSlotMigrationJobInProgress(job)) {
+        return;
+    }
+    int cross_node = 0;
+    clusterNode *n = getClusterNodeBySlotRanges(job->slot_ranges, &cross_node);
```

Comment:
make `cross_node` optional?
```suggestion
    clusterNode *n = getClusterNodeBySlotRanges(job->slot_ranges, NULL);
```

Reply:
Done